### ISSUE TITLE
Samsung Galaxy Book4 Pro 14" (NP940XGK) - speakers do not work

### ISSUE THREAD

**Issue by dgunay** [thesofproject/linux#5002]
EDIT: for anyone coming across this, here is the outcome of my attempt to reverse engineer the codec configuration: https://github.com/dgunay/galaxy-book4-pro-reverse-engineering

The speakers on the Galaxy Book4 Pro 14" NP940XGK don't work out of the box on Linux. The issue seems very similar to previously reported issues with Galaxy Book2 and Book3 laptop speakers not producing sound despite seeing the volume meter move when sound is being output. #4055 

I tried using many of the provided `hda-verb` scripts but they don't fix the problem - still no sound output.

Bluetooth headphones do work. I haven't tried using the onboard headphone jack yet.

My hunch is that the HDA verb sequence is different and needs to be recaptured for this specific device. I might give this a shot while I'm still in the return window, but it's my first time doing something like this so no promises.

**Comment by plbossart** [comment]
I presume you've looked at https://thesofproject.github.io/latest/getting_started/intel_debug/suggestions.html for the verb tables?

**Comment by dgunay** [comment]
> I presume you've looked at https://thesofproject.github.io/latest/getting_started/intel_debug/suggestions.html for the verb tables?

I looked at the `Reverse-engineer the Windows audio driver` section yesterday while rabbit-holing in that direction but not the rest of it. I'll go through it step by step tonight.

Thank you for the very prompt response.

**Comment by dgunay** [comment]
Ran `alsa-info.sh`. Results: http://alsa-project.org/db/?f=6105eef8dd841227908a445246a6b2582aff57da

**Comment by dgunay** [comment]
I followed the steps outlined in `Disable SOF on PCI/HDaudio devices to test audio playback`. No sounds were produced, again despite the volume monitor showing movement on the output device:

![image](https://github.com/thesofproject/linux/assets/16074091/1f2fdb53-24cd-478e-bca4-8c374d3b8111)

**Comment by dgunay** [comment]
Followed the `Try booting into Windows first, then reboot into Linux` step. No dice. Sound works in Windows, continues to not work in Linux. I didn't expect this to work since I booted to Windows quite a bit at first to get firmware updates, but worth a shot.

**Comment by dgunay** [comment]
I'm going to save the reverse engineering step for later since the other steps look like much lower hanging fruit.

**Comment by dgunay** [comment]
For `Make sure the ME is enabled` - I'm not seeing any way of configuring ME in the BIOS for this laptop, and I think I remember Windows Update downloading drivers for Intel ME when I first booted the thing, so I'm going to just assume for now that I have ME and it is enabled.

**Comment by dgunay** [comment]
Tried `Test at the ALSA ‘hw’ device level`

I'm not totally sure what to make of the commands mentioned in the guide.

Installing the suggested packages didn't do anything. When I enabled `pulseaudio`, it actually broke audio in a different way (YouTube videos would no longer play in Firefox until I disabled it). Installing `pulseaudio-alsa` fixed that, but did not result in sound coming out of the speakers.

**Comment by dgunay** [comment]
I've done `Verify mixer settings` multiple times now so I'm pretty sure the mixer isn't muting my speakers.

**Comment by dgunay** [comment]
Enabled dynamic debugging and ran `sudo dmesg | rg -i 'snd|sof|pci'`

Maybe someone more knowledgeable than me can glean something from this: https://gist.github.com/dgunay/0eea4dc528c55d30906429e9419d8264

**Comment by dgunay** [comment]
I'm not really qualified to tell whether the SOF tracing step is necessary yet, but I have `sof-logger` installed now.

**Comment by dgunay** [comment]
I don't think `Digital mic issues` applies to me, so I'm skipping that step.

**Comment by dgunay** [comment]
Looks like my system does not have `ES8336 support` according to the output of alsa-info - skipping that step.

**Comment by dgunay** [comment]
I've now tried everything in the suggestions article to the best of my ability, except for the reverse engineering step. Don't know if I'll have the time to attempt that tonight, we'll see.

**Comment by dgunay** [comment]
I've run through the reverse engineering steps from https://asus-linux.org/blog/sound-2021-01-11/ now but haven't had any luck. Here are my realtek driver dumps from both windows and linux: https://gist.github.com/dgunay/3c9a326e7f653a6cb1c7d282bb41853f

I ultimately intend to return the laptop so I'm not going to be able to keep grinding this problem down. I'm hoping this at least saves someone some time down the road, or perhaps someone more knowledgeable can use this to make a patch. 

Wasn't able to figure out how to reconfigure the pins and HDA verbs. I'll be cleaning up my work and pushing up a repository soon.

**Comment by dgunay** [comment]
Repository with all the dumps and some accompanying information: https://github.com/dgunay/galaxy-book4-pro-reverse-engineering

**Comment by mjkim0727** [comment]
Is this verb work on NT961XGK-K02/C?

I use Galaxy Book 4 Pro 16" South Korea Domestic model.

**Comment by dgunay** [comment]
> Is this verb work on NT961XGK-K02/C?
> 
> I use Galaxy Book 4 Pro 16" South Korea Domestic model.

I wasn't able to get it to work for me and I don't have the 16", but you're welcome to try it out.

**Comment by mayanksingh2298** [comment]
I use Galaxy book 3 pro 14"

I also can't get it to work using any list. Any help?

**Comment by mixim-mixim** [comment]
I also use Galaxy Book 4 Pro 16" (Europe version)
Tried everything mentioned here (all hadd-verb scripts)
nothing fixed, speakers always no sound, jack is ok

Please help !

**Comment by olwe1** [comment]
I have the same issue on Galaxy Book 4 Pro 14" (Europe version)

**Comment by mayanksingh2298** [comment]
Hey @mixim-mixim  and @olwe1,

I can confirm this list works for me: https://github.com/joshuagrisham/galaxy-book2-pro-linux/blob/main/sound/necessary-verbs.sh

I just tried this today.

Make sure your volume control looks something like this in your PulseAudioControl:
![image](https://github.com/thesofproject/linux/assets/25145295/29ee5cde-1803-4d95-b184-31e902f0ea2f)
(I was setting this to ProAudio and it didn't work then)


I'm on Galaxy Book 3 Pro 14":
![image](https://github.com/thesofproject/linux/assets/25145295/faf30ae1-b194-4459-b7bb-ee0ad3b68f1a)

**Comment by mixim-mixim** [comment]
> Hey @mixim-mixim and @olwe1,
> 
> I can confirm this list works for me: https://github.com/joshuagrisham/galaxy-book2-pro-linux/blob/main/sound/necessary-verbs.sh

Unfortunatly this does not work on a galaxy book4 ... at least not for mine :(

**Comment by PowerKiKi** [comment]
@mjkim0727 does your webcam work on your Galaxy Book 4 Pro 16" ? What linux are you using, Ubuntu ?

@mixim-mixim, what about your webcam, is it working ? which model of Galaxy Book 4 do you have ? pro/ultra 14"/16" ?

**Comment by mixim-mixim** [comment]
@PowerKiKi 
my webcam is also NOT working and my model is Galaxy Book 4 Pro 360 ... Fedora 40 (latest release) ... Kernel is 6.9.x

**Comment by PowerKiKi** [comment]
@mixim-mixim, thanks for the info. The webcam has been fixed very recently, according to https://github.com/intel/ipu6-drivers/pull/230. You might want to try that out...

**Comment by mjkim0727** [comment]
@PowerKiKi 

Webcam isn't working at Galaxy Book 4 Pro 16"

**Comment by PowerKiKi** [comment]
@mjkim0727, it should work now if you can use https://github.com/intel/ipu6-drivers/pull/230

**Comment by joshuagrisham** [comment]
Just realized maybe I should cross-post this here... I took a look at the RtHDDump file provided by @dgunay ..
If you want I can try to help  you get this working. We just finished submitting a patch that should be working for many GB2 and GB3 models and my guess is that it is quite similar for the GB4.
However based on the RtHDDump files it does not seem exactly the same (different coef indexes and possibly even different speaker amp IDs ??).

Is it possible for someone to run a trace using my version of QEMU from here?  
https://github.com/joshuagrisham/galaxy-book2-pro-linux/tree/main/sound/qemu

(more info here: https://github.com/thesofproject/linux/issues/4055#issuecomment-2323411911)

**Comment by mixim-mixim** [comment]
i took a look to ur script startvm.sh ... but i miss the 

qemu-system-x86_64

file ... where to get it from ?

**Comment by dginhoux** [comment]
Hi, 

i'm still, like everybody here trying and testing lot of things, but never to find a positive result.

It seem to be a big problem with all meteorlake laptop.
This king of problem remember me like wit 2.2 kernel and 56k "winmodems" lol.

**Comment by raelgc** [comment]
I have the Book 4 Ultra, same issue. I even tried the QemuHDADump. But it doesn't matter if Win10 or 11 VM, I could not get sound working (correct Intel drivers installed).

**Comment by dginhoux** [comment]
Read this about book, book2 and book3

https://github.com/thesofproject/linux/issues/4055#issuecomment-3007847242


I'm searching for a way to try different value but don't know and understand to do...

**Comment by stephanlensky** [comment]
I have a Galaxy Book 4 360 (NP960QGK) running Arch Linux and am also having problems. Like others, I ran Windows through Qemu with passed through PCI devices, but was still not able to get any speaker output.

I uploaded instructions for setting up the Windows VM in Arch Linux here: https://github.com/stephanlensky/galaxy-book4-linux-sound/tree/main/qemu

I didn't attempt any dumps from the VM yet since the drivers do not seem to be working properly there.

**Comment by dginhoux** [comment]
I think this computer is a really bad hardware/software and under linux, nothing will work... samsung say "fuck you with your linux, our computer are made for bill gates"

**Comment by BreadJS** [comment]
> I think this computer is a really bad hardware/software and under linux, nothing will work... samsung say "fuck you with your linux, our computer are made for bill gates"

I have a Galaxy Book 4 (NP940XGK-KG2NL). I am willing to run that QEMU version of yours to log the things that you need. Would you still want to do dumps with me? Could you also connect me on Discord with username BreadJS to collaborate on this issue (I am almost always available and my timezone is GMT+1)

**Comment by stephanlensky** [comment]
👋 As the bot mentioned, I've put a bounty on this issue! I'm especially interested in fixing it for my laptop (Galaxy Book 4 360 NP960QGK) but I believe the speaker configuration is likely the same on the regular Book 4 Pro.

For anyone interested in this, please feel free to reach out directly, you can find my mail on my personal site linked on my profile. I am happy to help run any tests required, or if you are a trusted maintainer I would also be willing to send my laptop to you. Thanks!

**Comment by stephanlensky** [comment]
Sorry for the double-post, I just wanted to share that an additional pledge of $500 USD has been posted for fixing the speakers of the related Galaxy Book 4 360 (NP960QGK) model.

I've included complete details in my repository here, as well as a bunch of log dumps from my laptop: https://github.com/stephanlensky/galaxy-book4-linux-sound

I'm going to create a separate issue in a moment specifically for the NP960QGK model in case the configuration does end up being different.

**Comment by stephanlensky** [comment]
@BreadJS has submitted a patch to fix this in #5616.

I have once again updated my repository here https://github.com/stephanlensky/galaxy-book4-linux-sound with instructions for applying the patch on Arch Linux while we wait for it to get merged. Thanks all 🙂

**Comment by dginhoux** [comment]
Congratulations ! @BreadJS 
I hope ear sound soon on my ubuntu setup !

**Comment by mateusluizfb** [comment]
Having same issue on my Samsung Book 4 Ultra. I'm using Fedora, will come back with an update after applying the patch as well

**Comment by Andycodeman** [comment]
For anyone looking for a workaround until this is merged upstream — I've packaged the fix from PR #5616 as a DKMS installer that works without recompiling the kernel:

https://github.com/Andycodeman/samsung-galaxy-book4-linux-fixes/tree/main/speaker-fix

It builds the MAX98390 HDA side-codec driver out-of-tree, handles I2C device creation via systemd, and auto-removes itself when native kernel support lands. Tested on Galaxy Book4 Ultra, Ubuntu 24.04, kernel 6.17.

**Comment by dginhoux** [comment]
Awesome job @Andycodeman just tested on the Pro4 and sound on the left side only

**Comment by Andycodeman** [comment]
> Awesome job [@Andycodeman](https://github.com/Andycodeman) just tested on the Pro4 and sound on the left side only

Interesting - I'm on Ultra4 but comes out of both. I'll take a look.

If you get a chance, can you run these and let me know the results...

  # Check which amps actually probed                                                                                                                                                                                                           
  journalctl -k -b | grep "MAX98390 HDA I2C probe"

  # Check which I2C devices exist
  ls /sys/bus/i2c/devices/ | grep -E '003[89cd]|MAX'

  # Check for driver errors
  journalctl -k -b | grep -i "max98390.*fail\|max98390.*err"

**Comment by Santigf12** [comment]
Hey @Andycodeman, awesome work. I was able to get to work in Fedora too (Galaxy Book4 Ultra model NP960XGL-XG3CA), Kernel 6.18.9 Fedora 43

The only extra thing I had to do was to sign the modules manually, so instead of the script’s MOK key path, I signed the DKMS-built modules using Fedora’s already-enrolled key (akmods/MOK). After that, everything loaded fine and the setup service brought up all 4 amps.

Really appreciate this, have a great day.

**Comment by Andycodeman** [comment]
Glad to hear it @Santigf12  and thanks for the update with your specific adjustment! I'll look to add notes and streamline more.

Edit: Added specific support for signing on Fedora so hopefully Fedora users won't have to do this manually like you did.

**Comment by dginhoux** [comment]
Hi @Andycodeman, 

Thank for your help, this is what i have : 
```
root@pc-dg21-u:~# journalctl -k -b | grep "MAX98390 HDA I2C probe"
Feb 14 14:02:20 pc-dg21-u kernel: max98390-hda i2c-MAX98390:00: MAX98390 HDA I2C probe: addr=0x38 index=0 name=i2c-MAX98390:00
root@pc-dg21-u:~# ls /sys/bus/i2c/devices/ | grep -E '003[89cd]|MAX'
i2c-MAX98390:00
root@pc-dg21-u:~# journalctl -k -b | grep -i "max98390.*fail|max98390.*err"
```

**Comment by dginhoux** [comment]
Just pulled yours repos with all yours commits of the day and test a fresh ubuntu 25.10 with latest mainline kernel 6.18.7.
Uninstall and install again, then reboot ; secureboot is enabled and working fine and try without it too (same result)

[diag-sgb4p_1.log](https://github.com/user-attachments/files/25318305/diag-sgb4p_1.log)

**Comment by Andycodeman** [comment]
> Just pulled yours repos with all yours commits of the day and test a fresh ubuntu 25.10 with latest mainline kernel 6.18.7. Uninstall and install again, then reboot ; secureboot is enabled and working fine and try without it too (same result)
> 
> [diag-sgb4p_1.log](https://github.com/user-attachments/files/25318305/diag-sgb4p_1.log)

Thanks for the logs!  Helped find the bug — the I2C bus detection was failing on your model. The setup script assumed a specific sysfs path structure that doesn't match the Pro4.
                                                                                                                                                                                                                                               
Your service log confirms it:                                      
max98390-hda: No MAX98390 ACPI device found on I2C bus        

This means only the first amp (0x38, left woofer) was active — the other 3 were never created. Just pushed a fix.
Get the latest from the repo and run the install script again.

After reboot, verify all amps probed:
journalctl -k -b | grep "MAX98390 HDA I2C probe"
ls /sys/bus/i2c/devices/ | grep -E '003[89cd]|MAX'

You should see 4 amps (0x38, 0x39, 0x3c, 0x3d) and audio from both sides.

**Comment by dginhoux** [comment]
Hi, i'm just trying to find and understand the same things by playing with your find_i2c_bus, find_present_addrs and the content of my device /sys content : 

```
dginhoux@pc-dg21-u:/sys/bus/acpi/devices/MAX98390:00$ ll
total 0
drwxr-xr-x 4 root root    0 Feb 14 22:13 ./
drwxr-xr-x 7 root root    0 Feb 14 22:30 ../
-r--r--r-- 1 root root 4096 Feb 14 22:30 hid
-r--r--r-- 1 root root 4096 Feb 14 22:30 modalias
-r--r--r-- 1 root root 4096 Feb 14 22:30 path
lrwxrwxrwx 1 root root    0 Feb 14 22:11 physical_node@ -> ../../../../../pci0000:00/0000:00:15.2/i2c_designware.2/i2c-2/i2c-MAX98390:00
drwxr-xr-x 2 root root    0 Feb 14 22:30 power/
-r--r--r-- 1 root root 4096 Feb 14 22:30 status
lrwxrwxrwx 1 root root    0 Feb 14 22:02 subsystem@ -> ../../../../../../bus/acpi
-rw-r--r-- 1 root root 4096 Feb 14 22:30 uevent
-r--r--r-- 1 root root 4096 Feb 14 22:30 uid
drwxr-xr-x 3 root root    0 Feb 14 22:30 wakeup/



dginhoux@pc-dg21-u:/sys/bus/i2c/devices$ ll
total 0
drwxr-xr-x 2 root root 0 Feb 14 22:02 ./
drwxr-xr-x 4 root root 0 Feb 14 22:02 ../
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-0@ -> ../../../devices/pci0000:00/0000:00:15.0/i2c_designware.0/i2c-0
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-1@ -> ../../../devices/pci0000:00/0000:00:15.1/i2c_designware.1/i2c-1
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-10@ -> ../../../devices/pci0000:00/0000:00:02.0/i2c-10
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-11@ -> ../../../devices/pci0000:00/0000:00:02.0/i2c-11
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-12@ -> ../../../devices/pci0000:00/0000:00:02.0/i2c-12
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-13@ -> ../../../devices/pci0000:00/0000:00:02.0/i2c-13
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-14@ -> ../../../devices/pci0000:00/0000:00:02.0/drm/card1/card1-eDP-1/i2c-14
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-15@ -> ../../../devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-1/i2c-15
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-16@ -> ../../../devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-2/i2c-16
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-17@ -> ../../../devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-3/i2c-17
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-2@ -> ../../../devices/pci0000:00/0000:00:15.2/i2c_designware.2/i2c-2
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-3@ -> ../../../devices/pci0000:00/0000:00:15.3/i2c_designware.3/i2c-3
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-4@ -> ../../../devices/pci0000:00/0000:00:1f.4/i2c-4
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-5@ -> ../../../devices/pci0000:00/0000:00:02.0/i2c-5
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-6@ -> ../../../devices/pci0000:00/0000:00:02.0/i2c-6
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-7@ -> ../../../devices/pci0000:00/0000:00:02.0/i2c-7
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-8@ -> ../../../devices/pci0000:00/0000:00:02.0/i2c-8
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-9@ -> ../../../devices/pci0000:00/0000:00:02.0/i2c-9
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-GXTP7936:00@ -> ../../../devices/pci0000:00/0000:00:15.1/i2c_designware.1/i2c-1/i2c-GXTP7936:00
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-MAX98390:00@ -> ../../../devices/pci0000:00/0000:00:15.2/i2c_designware.2/i2c-2/i2c-MAX98390:00
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-OVTI02C1:00@ -> ../../../devices/pci0000:00/0000:00:15.3/i2c_designware.3/i2c-3/i2c-OVTI02C1:00
lrwxrwxrwx 1 root root 0 Feb 14 22:02 i2c-ZNT0001:00@ -> ../../../devices/pci0000:00/0000:00:15.0/i2c_designware.0/i2c-0/i2c-ZNT0001:00
```

**Comment by dginhoux** [comment]
Yes, and now it's fully working left/right with your latest commit (dd30c15).
I hope your work will help lot of samsung laptop owners.
Amazing work
```
dginhoux@pc-dg21-u:~/samsung-galaxy-book4-linux-fixes/speaker-fix$ ./max98390-hda-i2c-setup.sh
max98390-hda: No additional amplifiers found on bus 2 (1-amp or already bound)




inhoux@pc-dg21-u:~/samsung-galaxy-book4-linux-fixes/speaker-fix$  journalctl -k -b | grep "MAX98390 HDA I2C probe"
Feb 14 22:02:28 pc-dg21-u kernel: max98390-hda i2c-MAX98390:00: MAX98390 HDA I2C probe: addr=0x38 index=0 name=i2c-MAX98390:00
Feb 14 22:38:32 pc-dg21-u kernel: max98390-hda i2c-MAX98390:00: MAX98390 HDA I2C probe: addr=0x38 index=0 name=i2c-MAX98390:00
Feb 14 22:38:33 pc-dg21-u kernel: max98390-hda 2-0039: MAX98390 HDA I2C probe: addr=0x39 index=1 name=2-0039
Feb 14 22:38:33 pc-dg21-u kernel: max98390-hda 2-003c: MAX98390 HDA I2C probe: addr=0x3c index=2 name=2-003c
Feb 14 22:38:33 pc-dg21-u kernel: max98390-hda 2-003d: MAX98390 HDA I2C probe: addr=0x3d index=3 name=2-003d



dginhoux@pc-dg21-u:~/samsung-galaxy-book4-linux-fixes/speaker-fix$  ls /sys/bus/i2c/devices/ | grep -E '003[89cd]|MAX'
2-0039
2-003c
2-003d
i2c-MAX98390:00



dginhoux@pc-dg21-u:~/samsung-galaxy-book4-linux-fixes/speaker-fix$  systemctl status max98390-hda-i2c-setup.service
● max98390-hda-i2c-setup.service - Create I2C devices for MAX98390 HDA speaker amplifiers
     Loaded: loaded (/etc/systemd/system/max98390-hda-i2c-setup.service; enabled; preset: enabled)
     Active: active (exited) since Sat 2026-02-14 22:38:32 CET; 8min ago
 Invocation: a3c277aebaae487c9f6625d19430b4d4
    Process: 20635 ExecStart=/usr/local/sbin/max98390-hda-i2c-setup.sh start (code=exited, status=0/SUCCESS)
   Main PID: 20635 (code=exited, status=0/SUCCESS)
   Mem peak: 3.7M
        CPU: 35ms

Feb 14 22:38:32 pc-dg21-u systemd[1]: Starting max98390-hda-i2c-setup.service - Create I2C devices for MAX98390 HDA speaker amplifiers...
Feb 14 22:38:32 pc-dg21-u max98390-hda-i2c-setup.sh[20635]: max98390-hda: Found 3 additional amplifier(s) on bus 2: 0x39 0x3c 0x3d
Feb 14 22:38:32 pc-dg21-u systemd[1]: Finished max98390-hda-i2c-setup.service - Create I2C devices for MAX98390 HDA speaker amplifiers.


```

**Comment by Andycodeman** [comment]
@dginhoux - Awesome!  Glad to hear it worked as I thought it might have been a race condition but nope, looks like the last fix will work for book4 Pro users.  Thanks for the willingness to troubleshoot. :)

**Comment by dginhoux** [comment]
Because i'm in testing mode, i'll reinstall from scratch an ubuntu 25.10, just os upgrade and add start the patch.
Let me 20/30min...

Seen on the windows (dualboot) an new bios v13, i'm doing upgrade

**Comment by dginhoux** [comment]
Fully working on fresh ubuntu install 25.10, kernek 6.17.0 on the book4 pro14 ; and the same with mainline 6.18.7 kernel.

**Comment by stephanlensky** [comment]
@Andycodeman thanks for putting this together.

> Battery Impact: This workaround keeps the speaker amplifiers powered on at all times (even when not playing audio), which uses an estimated 0.3–0.5W extra (~3–5% battery life). This is a limitation of the workaround approach

Is this a limitation just for your DKMS module, or this the case for the kernel patch in https://github.com/thesofproject/linux/pull/5616 as well?

Is there anyone on this thread who has experience contributing to the kernel? It would be great to get the code from https://github.com/thesofproject/linux/pull/5616 merged into mainline, but I'm not sure where to start.

**Comment by Andycodeman** [comment]
> [@Andycodeman](https://github.com/Andycodeman) thanks for putting this together.
> 
> > Battery Impact: This workaround keeps the speaker amplifiers powered on at all times (even when not playing audio), which uses an estimated 0.3–0.5W extra (~3–5% battery life). This is a limitation of the workaround approach
> 
> Is this a limitation just for your DKMS module, or this the case for the kernel patch in [#5616](https://github.com/thesofproject/linux/pull/5616) as well?
> 
> Is there anyone on this thread who has experience contributing to the kernel? It would be great to get the code from [#5616](https://github.com/thesofproject/linux/pull/5616) merged into mainline, but I'm not sure where to start.

  Good question! The always-on battery impact is only a limitation of this DKMS workaround only, not of the upstream PR #5616.

  The difference: the upstream PR modifies patch_realtek.c to add a Samsung-specific quirk that creates an HDA component master with proper playback hooks - the amps power on when audio plays and power off when it stops. This DKMS package
  can't modify patch_realtek.c (that would require recompiling the kernel), so we bypass the quirk entirely and set the amps to "always on" during driver probe instead. That's the trade-off for easy DKMS install vs. full kernel compilation (this DKMS does it intentionally to avoid kernel compilation).

  Once #5616 lands in mainline (expected kernel 7.0 or 7.1), proper power management comes with it — and this DKMS package will auto-detect native support and remove itself.

**Comment by stephanlensky** [comment]
> Once https://github.com/thesofproject/linux/pull/5616 lands in mainline (expected kernel 7.0 or 7.1),

Is there a patch already submitted for this? Or is it an LLM hallucination?

**Comment by Andycodeman** [comment]
> > Once [#5616](https://github.com/thesofproject/linux/pull/5616) lands in mainline (expected kernel 7.0 or 7.1),
> 
> Is there a patch already submitted for this? Or is it an LLM hallucination?

Definitely an estimate from AI due to submission date and number of responses, but no confirmation/guarantees on any of it.

**Comment by stephanlensky** [comment]
Well, we need to submit a patch via email unfortunately, as discussed a bit in https://github.com/thesofproject/linux/pull/5616. That pull request will never be merged since it's not touching code in SOF.

**Comment by Andycodeman** [comment]
> Well, we need to submit a patch via email unfortunately, as discussed a bit in [#5616](https://github.com/thesofproject/linux/pull/5616). That pull request will never be merged since it's not touching code in SOF.

Ah, didn't realize that was the current state - so the "real" fix is even further out as of this moment. :/
